File: _work/backend-watchdog-full-repair-02/src/iii-sandbox/test/sdk/sandbox.test.ts
import { describe, it, expect, vi, beforeEach } from "vitest";
import { Sandbox } from "../../packages/sdk/src/sandbox.js";
import { HttpClient } from "../../packages/sdk/src/client.js";

describe("Sandbox", () => {
  let mockClient: HttpClient;
  const info = {
    id: "sbx_test123",
    name: "test",
    image: "python:3.12-slim",
    status: "running" as const,
    createdAt: Date.now(),
    expiresAt: Date.now() + 3600000,
  };

  beforeEach(() => {
    mockClient = {
      get: vi.fn(),
      post: vi.fn(),
      del: vi.fn(),
      stream: vi.fn(),
    } as unknown as HttpClient;
  });

  describe("constructor and properties", () => {
    it("exposes id from info", () => {
      const sbx = new Sandbox(mockClient, info);
      expect(sbx.id).toBe("sbx_test123");
    });

    it("exposes status from info", () => {
      const sbx = new Sandbox(mockClient, info);
      expect(sbx.status).toBe("running");
    });

    it("creates filesystem instance", () => {
      const sbx = new Sandbox(mockClient, info);
      expect(sbx.filesystem).toBeDefined();
    });

    it("creates interpreter instance", () => {
      const sbx = new Sandbox(mockClient, info);
      expect(sbx.interpreter).toBeDefined();
    });

    it("info is mutable for refresh", () => {
      const sbx = new Sandbox(mockClient, info);
      sbx.info.status = "paused";
      expect(sbx.status).toBe("paused");
    });
  });

  describe("exec", () => {
    it("calls correct endpoint", async () => {
      const expected = {
        exitCode: 0,
        stdout: "hello",
        stderr: "",
        duration: 100,
      };
      (mockClient.post as any).mockResolvedValue(expected);
      const sbx = new Sandbox(mockClient, info);

      const result = await sbx.exec("echo hello");

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/exec",
        { command: "echo hello", timeout: undefined },
      );
      expect(result).toEqual(expected);
    });

    it("passes timeout when provided", async () => {
      (mockClient.post as any).mockResolvedValue({
        exitCode: 0,
        stdout: "",
        stderr: "",
        duration: 50,
      });
      const sbx = new Sandbox(mockClient, info);

      await sbx.exec("sleep 5", 10);

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/exec",
        { command: "sleep 5", timeout: 10 },
      );
    });

    it("propagates errors from client", async () => {
      (mockClient.post as any).mockRejectedValue(new Error("POST failed: 500"));
      const sbx = new Sandbox(mockClient, info);

      await expect(sbx.exec("bad")).rejects.toThrow("POST failed: 500");
    });
  });

  describe("execStream", () => {
    it("calls stream endpoint and returns async generator", async () => {
      const mockGen = (async function* () {
        yield '{"type":"stdout","data":"hello","timestamp":1}';
        yield '{"type":"exit","data":"0","timestamp":2}';
      })();
      (mockClient.stream as any).mockReturnValue(mockGen);
      const sbx = new Sandbox(mockClient, info);

      const gen = await sbx.execStream("echo hello");
      const chunks = [];
      for await (const chunk of gen) chunks.push(chunk);

      expect(mockClient.stream).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/exec/stream",
        { command: "echo hello" },
      );
      expect(chunks).toHaveLength(2);
      expect(chunks[0].type).toBe("stdout");
      expect(chunks[1].type).toBe("exit");
    });
  });

  describe("lifecycle operations", () => {
    it("pause calls correct endpoint", async () => {
      (mockClient.post as any).mockResolvedValue(undefined);
      const sbx = new Sandbox(mockClient, info);

      await sbx.pause();

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/pause",
      );
    });

    it("resume calls correct endpoint", async () => {
      (mockClient.post as any).mockResolvedValue(undefined);
      const sbx = new Sandbox(mockClient, info);

      await sbx.resume();

      expect(mockClient.post).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/resume",
      );
    });

    it("kill calls DELETE endpoint", async () => {
      (mockClient.del as any).mockResolvedValue(undefined);
      const sbx = new Sandbox(mockClient, info);

      await sbx.kill();

      expect(mockClient.del).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123",
      );
    });
  });

  describe("metrics", () => {
    it("calls metrics endpoint", async () => {
      const metrics = {
        sandboxId: "sbx_test123",
        cpuPercent: 5.2,
        memoryUsageMb: 128,
        memoryLimitMb: 512,
        networkRxBytes: 0,
        networkTxBytes: 0,
        pids: 3,
      };
      (mockClient.get as any).mockResolvedValue(metrics);
      const sbx = new Sandbox(mockClient, info);

      const result = await sbx.metrics();

      expect(mockClient.get).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123/metrics",
      );
      expect(result).toEqual(metrics);
    });
  });

  describe("refresh", () => {
    it("updates info from server", async () => {
      const updated = { ...info, status: "paused" as const };
      (mockClient.get as any).mockResolvedValue(updated);
      const sbx = new Sandbox(mockClient, info);

      const result = await sbx.refresh();

      expect(mockClient.get).toHaveBeenCalledWith(
        "/sandbox/sandboxes/sbx_test123",
      );
      expect(result.status).toBe("paused");
      expect(sbx.status).toBe("paused");
    });
  });
});

File: Core kernel Yaiwes/Backend watchdog workflow adaptativo/sandbox-runtime/iii-sandbox/packages/sdk-rust/tests/sandbox_test.rs
use std::sync::Arc;

use iii_sandbox::{ClientConfig, HttpClient, Sandbox, SandboxInfo};

fn make_sandbox(url: &str) -> Sandbox {
    let client = Arc::new(HttpClient::new(ClientConfig {
        base_url: url.to_string(),
        token: None,
        timeout_ms: None,
    }).unwrap());
    let info = SandboxInfo {
        id: "sbx-1".into(),
        name: "test".into(),
        image: "python:3.12-slim".into(),
        status: "running".into(),
        created_at: 1000,
        expires_at: 2000,
    };
    Sandbox::new(client, info)
}

#[tokio::test]
async fn test_exec() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/exec")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "exitCode": 0,
                "stdout": "hello\n",
                "stderr": "",
                "duration": 0.05
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let result = sbx.exec("echo hello", None).await.unwrap();
    assert_eq!(result.exit_code, 0);
    assert_eq!(result.stdout, "hello\n");
    assert_eq!(result.stderr, "");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_exec_with_timeout() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/exec")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "exitCode": 0,
                "stdout": "done",
                "stderr": "",
                "duration": 1.2
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let result = sbx.exec("sleep 1 && echo done", Some(5000)).await.unwrap();
    assert_eq!(result.exit_code, 0);
    assert_eq!(result.stdout, "done");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_clone_sandbox_with_name() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/clone")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "sbx-2",
                "name": "cloned",
                "image": "python:3.12-slim",
                "status": "running",
                "createdAt": 1000,
                "expiresAt": 2000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let cloned = sbx.clone_sandbox(Some("cloned")).await.unwrap();
    assert_eq!(cloned.id, "sbx-2");
    assert_eq!(cloned.name, "cloned");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_clone_sandbox_without_name() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/clone")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "sbx-3",
                "name": "auto-clone",
                "image": "python:3.12-slim",
                "status": "running",
                "createdAt": 1000,
                "expiresAt": 2000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let cloned = sbx.clone_sandbox(None).await.unwrap();
    assert_eq!(cloned.id, "sbx-3");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_pause() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/pause")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body("{}")
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    sbx.pause().await.unwrap();
    mock.assert_async().await;
}

#[tokio::test]
async fn test_resume() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/resume")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body("{}")
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    sbx.resume().await.unwrap();
    mock.assert_async().await;
}

#[tokio::test]
async fn test_kill() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("DELETE", "/sandbox/sandboxes/sbx-1")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body("{}")
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    sbx.kill().await.unwrap();
    mock.assert_async().await;
}

#[tokio::test]
async fn test_metrics() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/sandboxes/sbx-1/metrics")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "sandboxId": "sbx-1",
                "cpuPercent": 12.5,
                "memoryUsageMb": 128.0,
                "memoryLimitMb": 512.0,
                "networkRxBytes": 1024,
                "networkTxBytes": 2048,
                "pids": 5
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let metrics = sbx.metrics().await.unwrap();
    assert_eq!(metrics.sandbox_id, "sbx-1");
    assert_eq!(metrics.cpu_percent, 12.5);
    assert_eq!(metrics.memory_usage_mb, 128.0);
    assert_eq!(metrics.pids, 5);
    mock.assert_async().await;
}

#[tokio::test]
async fn test_snapshot_with_name() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/snapshots")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "snap-1",
                "sandboxId": "sbx-1",
                "name": "my-snap",
                "imageId": "sha256:abc",
                "size": 1024,
                "createdAt": 1000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let snap = sbx.snapshot(Some("my-snap")).await.unwrap();
    assert_eq!(snap.id, "snap-1");
    assert_eq!(snap.name, "my-snap");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_snapshot_without_name() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/snapshots")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "snap-2",
                "sandboxId": "sbx-1",
                "name": "auto",
                "imageId": "sha256:def",
                "size": 2048,
                "createdAt": 1000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let snap = sbx.snapshot(None).await.unwrap();
    assert_eq!(snap.id, "snap-2");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_restore() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("POST", "/sandbox/sandboxes/sbx-1/snapshots/restore")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "sbx-1",
                "name": "test",
                "image": "python:3.12-slim",
                "status": "running",
                "createdAt": 1000,
                "expiresAt": 2000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let restored = sbx.restore("snap-1").await.unwrap();
    assert_eq!(restored.id, "sbx-1");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_list_snapshots() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/sandboxes/sbx-1/snapshots")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "snapshots": [
                    {
                        "id": "snap-1",
                        "sandboxId": "sbx-1",
                        "name": "s1",
                        "imageId": "sha256:aaa",
                        "size": 512,
                        "createdAt": 1000
                    },
                    {
                        "id": "snap-2",
                        "sandboxId": "sbx-1",
                        "name": "s2",
                        "imageId": "sha256:bbb",
                        "size": 1024,
                        "createdAt": 2000
                    }
                ]
            })
            .to_string(),
        )
        .create_async()
        .await;

    let sbx = make_sandbox(&server.url());
    let list = sbx.list_snapshots().await.unwrap();
    assert_eq!(list.snapshots.len(), 2);
    assert_eq!(list.snapshots[0].id, "snap-1");
    assert_eq!(list.snapshots[1].id, "snap-2");
    mock.assert_async().await;
}

#[tokio::test]
async fn test_refresh() {
    let mut server = mockito::Server::new_async().await;
    let mock = server
        .mock("GET", "/sandbox/sandboxes/sbx-1")
        .with_status(200)
        .with_header("content-type", "application/json")
        .with_body(
            serde_json::json!({
                "id": "sbx-1",
                "name": "test",
                "image": "python:3.12-slim",
                "status": "paused",
                "createdAt": 1000,
                "expiresAt": 2000
            })
            .to_string(),
        )
        .create_async()
        .await;

    let mut sbx = make_sandbox(&server.url());
    assert_eq!(sbx.status(), "running");
    let updated = sbx.refresh().await.unwrap();
    assert_eq!(updated.status, "paused");
    assert_eq!(sbx.status(), "paused");
    mock.assert_async().await;
}
